README.org
* Bangkok

Bangkok reads chess game descriptions and re-play the games. Notice of
events (moves, captures, checks, etc.) are sent to a listener. Bangkok comes
with a listener that generates a MIDI file. In other words, the chess game
is turned into music.

Bangkok originated as the code for an art project by Tom Peak
<himself@tompeak.com>.

The Web site of Bangkok is (http://github.com/jimm/bangkok). Bangkok is also
available as a RubyGem.


** Recent Changes

*** 0.1.2

- GameListener#move outputs two notes for a move and sets CC_PORTAMENTO_TIME
  so there is a glide from the first note to the second. The total length of
  the two ntoes is proportional to the distance the piece travels. The first
  note is short (a 32nd note), and the second takes up the rest of the time,
  thus allowing the portamento to have its effect.

- GameListener#move also outputs multiple volume and pan values, moving
  smoothly from the original to the new value for the duration of the move's
  notes.

- New Square.at method returns a ready-made Square. Since they're immutable,
  this reduces the number of objects that need to be created.

- More tests.

*** 0.1.1

- Fixed bugs in the bin/bangkok script.

- Implemented the -c command line argument.

- Fixed a bug in the GameListener that prevented the MIDI file from being
  output.

- Added parsing of the config file to GameListener.


** Dependencies

The MIDI generation portion of Bangkok requires midilib 0.8.4 or later. If you
install Bangkok as a RubyGem, it will fetch and install midilib for you.

The test suite in the tests directory requires the testing framework TestUnit,
which comes with Ruby 1.8 and later and can also be found in the Ruby
Application Archive (http://raa.ruby-lang.org).

To rebuild the gem or RDocs or run the tests easily, you can use the Rakefile
which requires Rake (http://rake.rubyforge.org).


** Installation

*** RubyGems Installation

To install Bangkok as a gem, type

#+BEGIN_SRC sh
  gem install bangkok
#+END_SRC

You may need root privileges to install the gem.

*** Manual Installation

After downloading and expanding the archive, you can install Bangkok with the
command

#+BEGIN_SRC sh
  ruby install.rb
#+END_SRC

(or)

#+BEGIN_SRC sh
  ruby install.rb --install-dir=my_directory
#+END_SRC

You may need root privileges to install Bangkok.

** Testing

#+BEGIN_SRC sh
  rake test
#+END_SRC

runs all of the tests in the test directory.


** Overview

Bangkok replays a chess game. It reads chess game .pgn files. Interesting
events during each game (moves, captures, checks, etc.) are sent to a
listener.

Bangkok comes with a GameListener that creates a MIDI sequence. The listener
is given to a ChessGame object, which then reads a .pgn file, creates a board,
gives the moves to the board so it can move the pieces, and tells the listener
when the game is over. At that time, the GameListener writes the MIDI sequence
out to a MIDI file.


** How to Use

*** From the command line

Here is how to generate a MIDI file from a chess match file, using the
built-in GameListener. (This command line is awkward, and needs to get
simpler.)

#+BEGIN_SRC sh
  bangkok [-c my_program_changes.rb] chess_game_file.pgn
#+END_SRC

Running that command creates the MIDI file chess_game_file.mid. The optional
configuration file my_program_changes.rb lets you change the default program
numbers used for each of the pieces. See examples/program_changes.rb for an
example file.

For example, to run bangkok on the chess game in the examples directory, type

#+BEGIN_SRC sh
  bangkok examples/game.pgn
#+END_SRC

*** Example Scripts and Files

Here are short descriptions of each of the files found in the examples
directory.

- examples/announcer.rb creates a new listener that "announces" the game by
  printing some descriptive text for everything that happens during the game.

- examples/game.pgn is an example chess match.

- examples/program_changes.rb is an example configuration file that shows you
  how to change the default program change values. The values in this file are
  the same as the default values used by GameListener.


** The Code

Only one class has anything to do with MIDI at all: GameListener. The rest of
the classes in Bangkok know only about chess moves. This separation allows you
to use Bangkok for your own nefarious chess-related purposes, whether musical
or un-.

GameListener creates a MIDI sequence from the chess game's moves. When a piece
moves, MIDI notes are generated and a few controller values are set. Any
listener you write must implement the methods
- =start_game(io)=
- =end_game=
- =move(piece, from_square, to_square)=
- =capture(attacker_piece, loser_piece)=
- =check=
- =checkmate=
- =pawn_to_queen(pawn)=

For each move, the GameListener generates three sets of MIDI events: one for
the move start, one for a point halfway between the start and the destination,
and one for the destination. GameListener#midi_for_position creates a pan
value (left = queenside, right = kingside), a volume value (max = center of
board, min = top and bottom of board), and a quarter note. The note is from a
C scale, where the low C = the color's home rank and the high C = the opposite
color's home rank.

#+BEGIN_EXAMPLE
  SPEAKER    midi 1 - 8    SPEAKER (black)

     low
      ^
      |
     high
    volume
      |
      v
     low

      <---- panning ---->

  SPEAKER    midi 9 - 16    SPEAKER (white)
#+END_EXAMPLE

ChessGame takes a listener in its constructor and creates a Board. The method
ChessGame#read_moves reads game data. Calling ChessGame#play moves the pieces,
which in turn sends messages to the listener so it can react. See bin/bangkok
and examples/announcer.rb for examples of how to use a ChessGame object.

Board creates pieces and hands moves to pieces. It tells the listener about
things like captures and castles.

A Square represents a location on a Board.

A Piece has a color and a Square. It tells the listener about its moves.
Subclasses of Piece such as Queen and Knight override
Piece#could_perform_move, which checks to see if that piece could move from
its current position to the move's destination square. could_perform_move is
called by the Board to determine which piece on the board needs to be moved.

A Move has a color, a destination square, an optional starting rank or file,
and optional modifiers such as check or capture. A castle Move has no
destination square (the Board moves the pieces).


** Resources

The Ruby Web site (http://www.ruby-lang.org/en/index.html) contains an
introduction to Ruby, the Ruby Application Archive (RAA) at
http://raa.ruby-lang.org, and pointers to more information.


_Programming Ruby, The Pragmatic Programmer's Guide_, by David Thomas
and Andrew Hunt, is a well-written and practical introduction to Ruby. Its Web
page at http://www.rubycentral.com/book also contains a wealth of Ruby
information. Though the book is available online, I encourage you to purchase
a copy of the second edition, available at
http://pragmaticprogrammer.com/titles/ruby/.

midilib's home is http://midilib.rubyforge.org.


* To Do

#+INCLUDE: "TODO.org"


* Support

- Visit the forums, bug list, and mailing list pages at
  http://rubyforge.org/projects/bangkok

- Send email to Jim Menard at mailto:jim@jimmenard.com

- Ask on the ruby-talk mailing list


* Administrivia

Author:: Jim Menard (mailto:jim@jimmenard.com)
Copyright:: Copyright (c) 2005 Jim Menard
License:: Distributed under the same license as Ruby.


** Copying

Bangkok is copyrighted free software by Jim Menard and is released under the
same license as Ruby. See the Ruby license at
http://www.ruby-lang.org/en/LICENSE.txt.

Bangkok may be freely copied in its entirety providing this notice, all
source code, all documentation, and all other files are included.

Bangkok is Copyright (c) 2005 by Jim Menard.


** Warranty

This software is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantability and fitness for a particular purpose.


ChangeLog
2013-10-28  Jim Menard  <jim@jimmenard.com>

	* lib/bangkok/square.rb (Square::initialize): updated for Ruby
	1.9+, including Ruby 2.0.
	* Version bumped up to 0.2.0.

2005-03-26  Jim Menard  <jimm@io.com>

	* lib/bangkok/gamelistener.rb (GameListener::end_game): call
	Track#sort, not Track#recalc_delta_from_times, because we now
	generate events out of order.
	(GameListener::move): generate data differently. Call new
	generate_* methods. Generate multiple volume and pan values based
	on distance traveled.
	(GameListener::interpolate): created.
	(GameListener::generate_notes): created.
	(GameListener::file_to_pan): created.
	(GameListener::rank_to_volume): created.
	(GameListener::generate_volume): created.
	(GameListener::generate_pan): created.
	(GameListener::generate_portamento): created.

	* lib/bangkok/square.rb (Square::distance_to): created.
	(Square::at): created.

	* lib/bangkok/piece.rb: use new Square.at method where appropriate.

	* lib/bangkok/board.rb: use new Square.at method where appropriate.

2005-03-25  Jim Menard  <jimm@io.com>

	* Version 0.1.1 released.

	* lib/bangkok/chessgame.rb: removed shebang line.

	* lib/bangkok/gamelistener.rb (GameListener::initialize): added
	config file path argument.
	(GameListener::read_config): created.
	(GameListener::end_game): io is an instance var, not a local var.
	Restructured code around filling in config information.

	* bin/bangkok: fixed really stupid syntax errors. Added -c command
	line argument. Pass config file path to GameListener.

	* Version 0.1.0 released.

2005-03-24  Jim Menard  <jimm@io.com>

	* Initial project creation.


lib/bangkok/gamelistener.rb
begin
  require 'midilib'
rescue LoadError
  require 'rubygems'
  require 'midilib'
end
include MIDI

# The pieces and board call methods on an instance of GameListener, which
# turns those events into MIDI events.
class GameListener

  PIECE_NAMES = {
    :P => "Pawn",
    :R => "Rook",
    :N => "Night",
    :B => "Bishop",
    :Q => "Queen",
    :K => "King"
  }

  # Build notes to play for ranks
  RANK_TO_NOTE = {}
  RANK_TO_NOTE[:white] = [64, 66, 68, 69, 71, 73, 75, 76]
  RANK_TO_NOTE[:black] = RANK_TO_NOTE[:white].reverse

  PIECE_MIDI_INFO = {}
  PIECE_MIDI_INFO[:white] = {}
  PIECE_MIDI_INFO[:black] = {}

  # Used by config files to set the program change value for a piece.
  def GameListener.white(piece_sym, program_change)
    PIECE_MIDI_INFO[:white][piece_sym] = program_change
  end

  # Used by config files to set the program change value for a piece.
  def GameListener.black(piece_sym, program_change)
    GameListener::PIECE_MIDI_INFO[:black][piece_sym] = program_change
  end

  # Fill in some default values, in case the user does not specify any new
  # ones.
  black :K, 0
  black :Q, 8
  black :R, 16
  black :B, 24
  black :N, 32
  black :P, 4
  white :K, 4
  white :Q, 12
  white :R, 20
  white :B, 28
  white :N, 36
  white :P, 44

  attr_reader :seq              # MIDI::Sequence

  def initialize(config_file_path=nil)
    read_config(config_file_path) if config_file_path
  end

  def read_config(config_file_path)
    IO.readlines(config_file_path).each { | line |
      line.chomp!
      class_eval(line)
    }
  end

  def track_of(piece)
    i = [:white, :black].index(piece.color) * 6 +
      [:P, :R, :N, :B, :Q, :K].index(piece.piece)
    @seq.tracks[i + 1]          # 0'th track is temp track
  end

  def channel_of(piece)
    [:white, :black].index(piece.color) * 8 +
      [:P, :R, :N, :B, :Q, :K].index(piece.piece)
  end

  # --
  # ================================================================
  # Listener interface
  # ================================================================
  # ++

  def start_game(io)
    @io = io
    @seq = Sequence.new
    track = Track.new(@seq)     # Tempo track
    track.name = "Tempo track"
    @seq.tracks << track

    @first_note_delta = @seq.note_to_delta('32nd')
    @portamento_start = @seq.note_to_delta('64th')
    @max_delta =
	@seq.length_to_delta(Square.at(0, 0).distance_to(Square.at(7, 7)))

    create_tracks()
    @time_from_start = 0
  end

  def end_game
    # When we created events, we set their start times, not their delta times.
    # Now is the time to fix that. Sort sorts by start times then calls
    # recalc_delta_from_times.
    @seq.tracks.each { | t | t.sort }
    @seq.write(@io)
  end

  # Move +piece+ +from+ one space +to+ another. Generate two notes and sets
  # CC_PORTAMENTO_TIME so there is a glide from the first note to the second.
  # Also output multiple volume and pan values, moving smoothly from the
  # original to the new value.
  def move(piece, from, to)
    raise "from #{from} may not be off the board" unless from.on_board?

    # Do nothing if the piece moves off the board, because either capture()
    # or pawn_to_queen() will be called.
    return unless to.on_board?

    track = track_of(piece)
    channel = channel_of(piece)

    dist = from.distance_to(to)
    total_delta = @seq.length_to_delta(dist) # quarter note per space

    generate_notes(track, channel, total_delta, piece, from, to)
    generate_portamento(track, channel, total_delta)

    steps = interpolate(16, 64, 0, @max_delta, total_delta).to_i
    delta = (total_delta / steps).to_i
    start = @time_from_start

    steps.times { | step |
      val = rank_to_volume(interpolate(from.rank, to.rank, 0, steps-1, step))
      generate_volume(track, channel, start, val)
      
      val = file_to_pan(interpolate(from.file, to.file, 0, steps-1, step))
      generate_pan(track, channel, start, val)

      start += delta
    }

    @time_from_start += total_delta
  end

  def capture(attacker, loser)
  end

  def check
  end

  def checkmate
  end

  def pawn_to_queen(pawn)
  end

  # --
  # ================================================================
  # End of listener interface
  # ================================================================
  # ++

  def create_tracks
    [:white, :black].each_with_index { | color, chan_base_offset |
      [:P, :R, :N, :B, :Q, :K].each_with_index { | piece_sym, chan_offset |
        track = Track.new(@seq)
        @seq.tracks << track

        track.name = "#{color.to_s.capitalize} #{PIECE_NAMES[piece_sym]}"

        program_num = PIECE_MIDI_INFO[color][piece_sym]
        track.instrument = GM_PATCH_NAMES[program_num]
        track.events << ProgramChange.new(chan_base_offset * 8 + chan_offset,
                                          program_num)
      }
    }
  end

  # Returns a value between range_min and range_max inclusive that is
  # proportional to +value+'s place between value_min and value_max. The
  # returned value may be floating point. If range_min and range_max or
  # value_min and value_max are out of order, they are swapped.
  def interpolate(range_min, range_max, value_min, value_max, value)
    range_min, range_max = range_max, range_min if range_min > range_max
    value_min, value_max = value_max, value_min if value_min > value_max
    return range_min if value == value_min
    frac = (value_max.to_f - value_min.to_f) / (value.to_f - value_min.to_f)
    return range_min + (range_max.to_f - range_min.to_f) / frac
  end

  # Generate two notes, one at the current start time that is only a 32nd note
  # long. The next follows immediately, and is for the +to+ square. Its length
  # is the remaining time.
  def generate_notes(track, channel, total_delta, piece, from, to)
    # First note is quickly followed by the second note
    note = RANK_TO_NOTE[piece.color][from.rank]
    e = NoteOnEvent.new(channel, note, 127)
    e.time_from_start = @time_from_start
    track.events << e

    e = NoteOffEvent.new(channel, note, 127)
    e.time_from_start = @time_from_start + @first_note_delta - 1
    track.events << e

    # Second note
    note = RANK_TO_NOTE[piece.color][to.rank]
    e = NoteOnEvent.new(channel, note, 127)
    e.time_from_start = @time_from_start + @first_note_delta
    track.events << e

    e = NoteOffEvent.new(channel, note, 127)
    e.time_from_start = @time_from_start + total_delta - 1
    track.events << e
  end

  # Translates a (possibly fractional) +file+ into an integer CC_PAN value.
  def file_to_pan(file)
    return interpolate(0, 127, 0, 7, file).to_i
  end

  # Translates a (possibly fractional) +rank+ into an integer CC_VOLUME value.
  def rank_to_volume(rank)
    rank = 3.5 - (3.5 - rank).abs
    return interpolate(10, 127, 0, 3.5, rank).to_i
  end

  # Generates a single volume event. #move_to calls this multiple times,
  # passing in new +delta+ and +value+ values.
  def generate_volume(track, channel, delta, value)
    raise "volume: bogus value #{value}" unless value >= 0 && value <= 127
    e = Controller.new(channel, CC_VOLUME, value)
    e.time_from_start = @time_from_start + delta
    track.events << e
  end

  # Generates a single pan event. #move_to calls this multiple times, passing
  # in new +delta+ and +value+ values.
  def generate_pan(track, channel, delta, value)
    raise "pan: bogus value #{value}" unless value >= 0 && value <= 127
    e = Controller.new(channel, CC_PAN, value)
    e.time_from_start = @time_from_start + delta
    track.events << e
  end

  # Generates a single portamento event.
  def generate_portamento(track, channel, total_delta)
    e = Controller.new(channel, CC_PORTAMENTO_TIME, 0)
    e.time_from_start = @time_from_start
    track.events << e

    value = interpolate(32, 100, 0, @max_delta, total_delta).to_i
    raise "portamento: bogus value #{value}" unless value >= 0 && value <= 127
    e = Controller.new(channel, CC_PORTAMENTO_TIME, value)
		       
    e.time_from_start = @time_from_start + @portamento_start
    track.events << e
  end

end
